Source/NSubstitute/Core/ICustomHandlers.cs

```diff
@@ -0,0 +1,16 @@
+﻿using System.Collections.Generic;
+
+namespace NSubstitute.Core
+{
+    /// <summary>
+    ///     Factory method which creates <see cref="ICallHandler" /> from the <see cref="ISubstituteState" />.
+    /// </summary>
+    public delegate ICallHandler CallHandlerFactory(ISubstituteState substituteState);
+
+    public interface ICustomHandlers
+    {
+        IEnumerable<ICallHandler> Handlers { get; }
+        IDictionary<object, object> HandlerDataStorage { get; }
```

Comment:
Current implementation is that each `CallHandler` is stateless and substitute state is present in the `state` object. Handlers are recreated each time the route is composed.
My thoughts were to have place in `state` where each call handler could store it's permanent data per substitute (e.g. cached results). Later it could read it from storage when call is dispatched. Alternatively. each handler will need to track it's own dictionary to map substitute to state. That is much harder (and consts in term of memory/performance) than to have a place inside state where it could put its data.

Later I changed implementation and now single instance of custom handler is created per substitute. However, that is rather a "weak agreement" that we will not create handler more than once per substitute.